Implement a Python client

honzasp opened the pull request:
Adds a Python package with a sqld client, analogous to the TypeScript and Rust clients.

honzasp commented:
After this PR gets merged, I will publish the 0.1.0 version to PyPI.

psarna commented:
Very nice to see pytest that works in both local and remote mode!

I have one optional suggestion - maybe the http tests should be skipped instead of failing if the server is unreachable for any reason - that way `pytest` will work out of the box for people that just check out the repository.

Here's a suggested patch:
```diff
diff --git a/pypackages/libsql-client/tests/conftest.py b/pypackages/libsql-client/tests/conftest.py
index e43fd7f..2a3c755 100644
--- a/pypackages/libsql-client/tests/conftest.py
+++ b/pypackages/libsql-client/tests/conftest.py
@@ -1,10 +1,20 @@
 import os
 import pytest
+import requests
+
+def is_alive(url):
+    try:
+        return requests.get(url + "/health").status_code == 200
+    except:
+        return False
 
 @pytest.fixture(params=["http", "file"])
 def url(request, tmp_path):
     if request.param == "http":
-        return os.getenv("LIBSQL_CLIENT_TEST_URL", "http://localhost:8080")
+        url = os.getenv("LIBSQL_CLIENT_TEST_URL", "http://localhost:8080")
+        if not is_alive(url):
+            pytest.skip(f"Skipping an HTTP test: server {url} is unreachable")
+        return url
     elif request.param == "file":
         return f"file://{tmp_path.absolute() / 'test.db'}"
     else:
```

Then, with `pytest -v`, people are also going to see this, if they haven't launched a local server:
```
[sarna@sarna-pc libsql-client]$ pytest -v
============================================================================ test session starts ============================================================================
platform linux -- Python 3.10.9, pytest-7.2.1, pluggy-1.0.0 -- /usr/bin/python
cachedir: .pytest_cache
rootdir: /home/sarna/repo/sqld/pypackages/libsql-client
plugins: asyncio-0.20.3
asyncio: mode=strict
collected 24 items                                                                                                                                                          

tests/test_client.py::test_execute[http] SKIPPED (Skipping an HTTP test: server http://localhost:8080 is unreachable)                                                 [  4%]
tests/test_client.py::test_execute[file] PASSED                                                                                                                       [  8%]
tests/test_client.py::test_batch[http] SKIPPED (Skipping an HTTP test: server http://localhost:8080 is unreachable)                                                   [ 12%]
tests/test_client.py::test_batch[file] PASSED                                                                                                                         [ 16%]
tests/test_client.py::test_error[http] SKIPPED (Skipping an HTTP test: server http://localhost:8080 is unreachable)                                                   [ 20%]
tests/test_client.py::test_error[file] PASSED                                                                                                                         [ 25%]
tests/test_params.py::test_basic_types[http] SKIPPED (Skipping an HTTP test: server http://localhost:8080 is unreachable)                                             [ 29%]
tests/test_params.py::test_basic_types[file] PASSED                                                                                                                   [ 33%]
tests/test_params.py::test_big_integer[http] SKIPPED (Skipping an HTTP test: server http://localhost:8080 is unreachable)                                             [ 37%]
tests/test_params.py::test_big_integer[file] PASSED                                                                                                                   [ 41%]
tests/test_params.py::test_blob[http] SKIPPED (Skipping an HTTP test: server http://localhost:8080 is unreachable)                                                    [ 45%]
tests/test_params.py::test_blob[file] PASSED                                                                                                                          [ 50%]
tests/test_result.py::test_basic_types[http] SKIPPED (Skipping an HTTP test: server http://localhost:8080 is unreachable)                                             [ 54%]
tests/test_result.py::test_basic_types[file] PASSED                                                                                                                   [ 58%]
tests/test_result.py::test_big_integer[http] SKIPPED (Skipping an HTTP test: server http://localhost:8080 is unreachable)                                             [ 62%]
tests/test_result.py::test_big_integer[file] PASSED                                                                                                                   [ 66%]
tests/test_result.py::test_blob[http] SKIPPED (Skipping an HTTP test: server http://localhost:8080 is unreachable)                                                    [ 70%]
tests/test_result.py::test_blob[file] XPASS                                                                                                                           [ 75%]
tests/test_result.py::test_columns[http] SKIPPED (Skipping an HTTP test: server http://localhost:8080 is unreachable)                                                 [ 79%]
tests/test_result.py::test_columns[file] PASSED                                                                                                                       [ 83%]
tests/test_result.py::test_rows[http] SKIPPED (Skipping an HTTP test: server http://localhost:8080 is unreachable)                                                    [ 87%]
tests/test_result.py::test_rows[file] PASSED                                                                                                                          [ 91%]
tests/test_result.py::test_row_repr[http] SKIPPED (Skipping an HTTP test: server http://localhost:8080 is unreachable)                                                [ 95%]
tests/test_result.py::test_row_repr[file] PASSED                                                                                                                      [100%]

================================================================= 11 passed, 12 skipped, 1 xpassed in 0.36s =================================================================
```

honzasp commented:
Good point Piotr, I adapted your solution and skip the tests when the `LIBSQL_CLIENT_TEST_URL` is not defined and we cannot reach the default URL. However, I think it's better to let the tests fail when the env variable is defined, because in this case the user explicitly wanted us to connect to that URL.

honzasp commented on pypackages/libsql-client/libsql_client/client.py:
```diff
@@ -0,0 +1,45 @@
+        else:
+            raise ValueError(f"Unsupported URL scheme: {parsed_url.scheme!r}")
+
+    async def execute(self, sql: str, params: Sequence["Value"] = ()) -> "ResultSet":
```
They are in quotes because the names are not available at runtime. In this case, I import them only when type checking to avoid the unnecessary import at runtime. In other cases, this is used to resolve cyclical dependencies or to let me reorder definitions.

honzasp commented on pypackages/libsql-client/libsql_client/http_driver.py:
```diff
@@ -0,0 +1,75 @@
+        async with await self._session.post("/", json=req_body) as resp:
+            resp.raise_for_status()
+
+            resp_body = await resp.json(content_type=None)
```
sqld does not set the `Content-Type` header in the response. Aiohttp checks the content type by default, so I need to pass `None` to disable this check.

honzasp commented on pypackages/libsql-client/README.md:
```diff
@@ -0,0 +1,60 @@
+
+## Getting started
+
+To get started, you need `sqld` running somewhere. Then you can install this package with:
```
I just copied this sentence from the `libsql-client` in TypeScript :) Will add some basic instructions.

honzasp commented:
Thanks for review! :) I addressed your comments and added some doc strings.

MarinPostma commented on pypackages/libsql-client/libsql_client/http_driver.py:
```diff
@@ -0,0 +1,75 @@
+    if isinstance(value_json, int) or isinstance(value_json, float):
+        return value_json
+    elif isinstance(value_json, str):
+        # TODO: this might be a base64-encoded blob!
```
Uhh this is a bug, blob should be encoded as an object. fixing this right now

MarinPostma replied:
https://github.com/libsql/sqld/pull/152 -> here

honzasp replied:
Done

CodingDoug commented:
This library exposes a `Client` object with `execute` and `batch` methods, whereas the JS and Rust libraries expose a `Connection` object with `execute` and `transaction` methods.  Can we update this to maintain consistency?

Also I am hoping this can support named bound parameters in addition to the positional parameters that exist now.

honzasp commented:
> This library exposes a `Client` object with `execute` and `batch` methods, whereas the JS and Rust libraries expose a `Connection` object with `execute` and `transaction` methods. Can we update this to maintain consistency?

- I added the `transaction()` method, which simply runs a `batch()` sandwitched between BEGIN and COMMIT, to faithfully replicate the Rust client.
- As I argued on Slack, `Connection` is not a good name. This does not correspond to a TCP connection or a database connection, because there is no state maintained between the `execute()`/`batch()` calls, in contrast to every other SQL client that I know. I would prefer to keep `Client` in the Python library, and rename the TypeScript and Rust libraries to match.

> Also I am hoping this can support named bound parameters in addition to the positional parameters that exist now.

- Good catch, fixed! :)

honzasp commented:
bors merge
Will it work? :)

honzasp commented:
sudo bors merge
